services/api/service.go

```diff
@@ -835,6 +924,71 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
 		if err != nil {
 			log.WithError(err).Error("failed to increment builder-stats after getPayload")
 		}
+
+		builderStatus, err := api.redis.GetBlockBuilderStatus(bidTrace.BuilderPubkey.String())
+		if err != nil {
+			log.WithError(err).Error("failed to get builder status")
+		}
+
+		// Check if the block was valid if the builder is OptimisticActive or OptimisticDemoted.
+		if builderStatus == common.OptimisticActive || builderStatus == common.OptimisticDemoted {
+			// Set to locked while we process the winning block.
+			api.setStatusByCollateralID(bidTrace.BuilderPubkey.String(), common.OptimisticLocked)
```

Comment:
What if `handleSubmitNewBlock` for the next slot is somehow called before `handleGetPayload` for the current slot executes this line code? For example, what if `api.db.SaveDeliveredPayload` (a few lines above) takes a really long time? There is a race condition here, right?

I think there's also a more fundamental problem: we can't rely on `handleGetPayload` to determine the winning block. The reason is that the validator could collude with the builder to get the payload out of band, without ever calling `handleGetPayload`. This could even happen without collusion when multiple relays forwarded the same header and the proposer gets the payload from another relay first.

Reply:
from discussions offline, we are going to set value for currentOptimisticSlot and only optimistically process blocks one slot at a time. this will avoid these tricky race conditions and make our design easier to reason about. it also allows us to get rid of the `OptimisticLocked` status, because we know while we are validating the block that won the slot `n` auction, if the optimistic processing begins for slot `n+1`, then we must have already processed all the blocks in slot `n`, so we can be sure that if the winning block ended up being invalid, that builder has been demoted and thus doesn't have an opportunity to build invalid blocks on invalid blocks.

Reply:
this is going to be one of the major design changes of the PR 3 :-)